### PR TITLE
Feat 001 auth

### PR DESCRIPTION
**Pull request by GPabis**

### CODE REVIEW

File: backend/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import express from 'express';
+import authRouter from './routes/auth';
+import { config } from 'dotenv';
```
Comment (ukaszcholewka) [inline backend/src/index.ts]: You don't need this. `.env` file is read by default (at least in node v16.3.0)

File: backend/src/middleware/auth.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import jwt from 'jsonwebtoken';
+import { config } from 'dotenv';
```
Comment (ukaszcholewka) [inline backend/src/middleware/auth.ts]: https://github.com/GPabis/project-management-app/pull/2/files#diff-c20200a666b149a045e365999f581db8e04687cfc9569bdf4c9477cab954d324R3

File: backend/src/middleware/auth.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import { Request, Response, NextFunction } from 'express';
+
+config();
+
+const verifyToken = (req: Request, res: Response, next: NextFunction) => {
+    const token = req.query.token || req.headers['x-access-token'] || req.body.token;
+
+    if (!token) {
+        return res.status(403).send('A token is required for authentication');
```
Comment (ukaszcholewka) [inline backend/src/middleware/auth.ts]: I think it's better to create enum with status codes in some `constans` file

`enum STATUS_CODES = {
    SUCCESS: 200
    ...
}`

File: backend/src/middleware/database.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import mongoose from 'mongoose';
+import { config } from 'dotenv';
+
+config();
+
+const { MD_HOST, MD_PORT, MD_USER, MD_PASSWORD } = process.env;
+
+const connect = async () => {
+    const url = `mongodb://${MD_USER}:${MD_PASSWORD}@${MD_HOST}:${MD_PORT}/${MD_USER}?synchronize=true`;
```
Comment (ukaszcholewka) [inline backend/src/middleware/database.ts]: I think, it will be better if this const would be above function declaration.

File: backend/src/models/user.model.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+const Schema = mongoose.Schema;
+
+const userShema = new Schema<UserModel>({
+    email: {
+        type: String,
+        required: true,
+        unique: true,
+    },
+    username: {
```
Comment (ukaszcholewka) [inline backend/src/models/user.model.ts:18]: Username doesn't have to be unique?

File: backend/src/routes/auth.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    password: string;
+}
+
+interface loginBody {
+    email: string;
+    password: string;
+}
+
+const findUserByEmail = async (email: string) => {
+    const oldUser = await User.findOne({ email });
```
Comment (ukaszcholewka) [inline backend/src/routes/auth.ts]: `existingUser` instead of `oldUser `?

File: backend/src/routes/auth.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    return oldUser ? oldUser : false;
+};
+
+router.post('/register', async (req, res) => {
+    try {
+        const { email, username, password }: registerBody = req.body;
+
+        if (!(email && username && password)) {
```
Comment (ukaszcholewka) [inline backend/src/routes/auth.ts]: Sooo,  ` email="" username="" password="" ` is fine

File: backend/src/routes/auth.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        const tokenKey = process.env.TOKEN_KEY;
+
+        if (!tokenKey) {
+            return res.status(500).send('Server Problem - plesase try again later');
+        }
+
+        const token = jwt.sign({ user_id: user._id, email }, tokenKey, {
+            expiresIn: '2h',
+        });
+        user.token = token;
```
Comment (ukaszcholewka) [inline backend/src/routes/auth.ts]: https://www.google.com/search?q=js+mutation+is+bad&oq=js+mutation+is+bad&aqs=chrome..69i57j0i10j0j0i10l5j0i324i395i492l2.8958j1j7&sourceid=chrome&ie=UTF-8

TLDR: Object mutation is considering a bad practice. You can do: 
```
user = {
    ...user,
    token
}
```
or create new variable: 

```
const response = {
    ...user,
    token
}
```

or just simply: 
```
res.status(201).json({
    ...user,
    token
})
```

File: backend/src/routes/auth.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+        if (!(email && password)) {
+            res.status(400).send('All input is required!');
+        }
+
+        await connect();
+
+        const user = await findUserByEmail(email);
+
+        const tokenKey = process.env.TOKEN_KEY;
```
Comment (ukaszcholewka) [inline backend/src/routes/auth.ts]: Why not put it on top-level? There is no reason to declare this const in function.

File: frontend/src/App.tsx [hunk 10]
```diff
@@ -1,5 +1,24 @@
+import Header from "./components/Header";
+import FormField from "./components/Register/FormField";
+import Input from "./components/Register/Input";
+import Label from "./components/Register/Label";
+import Submit from "./components/Register/Submit";
```
Comment (ukaszcholewka) [inline frontend/src/App.tsx]: It can be done wayyyy better. 
Basically you have lots of default exports in different files. Now in `/components/Register/` you need to create `intex.ts` file and export everything in it

```
export { default as Header } from './Header'
export { default as FormField} from './FormField'
...
```
then you can import it as:

`import { Header, FormField, Input, Label } from "./components/Register"`

File: frontend/src/App.tsx [hunk 11]
```diff
@@ -1,5 +1,24 @@
...
+
+
+
```
Comment (ukaszcholewka) [inline frontend/src/App.tsx]: Can we change function to arrow function?

const App: FC = () => {...}

File: frontend/src/components/Header/Logo.tsx [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import styled from "styled-components";
+
+
+const Logo = styled.a`
+ a{
```
Comment (ukaszcholewka) [inline frontend/src/components/Header/Logo.tsx]: missing space `a {`

File: frontend/src/components/Header/Logo.tsx [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+   cursor: pointer;
+   font-size: 2rem;
+   color: #fff;   
+   font-family: 'Lato';
+   font-weight: bold;
+   transition: all 0.3s;
+   text-decoration: none;
+   :hover{
```
Comment (ukaszcholewka) [inline frontend/src/components/Header/Logo.tsx]: Shouldn't it be `&:hover `?

File: frontend/src/components/Header/NavElement.tsx [hunk 14]
```diff
@@ -0,0 +1,21 @@
+
+import styled from "styled-components";
+
+
+const NavElement = styled.li`
+  font-size: 1.6rem;
+  padding: 0 1rem;
+  margin: 0;
+
+  a{
```
Comment (ukaszcholewka) [inline frontend/src/components/Header/NavElement.tsx]: missing space

File: frontend/src/components/Header/index.tsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+          <Nav>
+
+            <NavElement>
+              <NavLink to='/login'>Login</NavLink>
+            </NavElement>
+
+            <NavElement>
+              <NavLink to='/register'>Register</NavLink>
+            </NavElement>
+
```
Comment (ukaszcholewka) [inline frontend/src/components/Header/index.tsx]: Why so many new lines?

```
<>
    <ContainerLayout background={tertiaryColor}>
      <Container>
        <HeaderContainer>

          <Logo>
            <Link to='/'>Project Management</Link>
          </Logo>

          <Nav>
            <NavElement>
              <NavLink to='/login'>Login</NavLink>
            </NavElement>
            <NavElement>
              <NavLink to='/register'>Register</NavLink>
            </NavElement>
          </Nav>

        </HeaderContainer>
      </Container>
    </ContainerLayout>
    {children}
  </>
```

File: frontend/src/components/util/ContainerLayout.tsx [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import styled from "styled-components";
+
+interface LayoutProps {
+  background?: string,
+}
+
+const ContainerLayout = styled.div<LayoutProps>`
+  background: ${props => props.background ? props.background : '#fffff'};
```
Comment (ukaszcholewka) [inline frontend/src/components/util/ContainerLayout.tsx]: `background: ${({ background }) => background ? background  : '#fff'}`

File: frontend/src/components/util/GlobalStyles.tsx [hunk 17]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (ukaszcholewka) [inline frontend/src/components/util/GlobalStyles.tsx]: What are those empty lines?

File: frontend/src/components/Header/index.tsx [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+const Header:FC = ({children}) => {
+  return (
+  <>
+    <ContainerLayout background={tertiaryColor}>
+      <Container>
+        <HeaderContainer>
+
+          <Logo>
+            <Link to='/'>Project Management</Link>
```
Comment (ukaszcholewka) [inline frontend/src/components/Header/index.tsx]: `<Logo />` renders `<a />` element, also `<Link />` renders `<a  />` element. 
By the standard of w3c you can not have `a` tag inside of another `a` tag

https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element

**Review by ukaszcholewka — CHANGES_REQUESTED** [review]
Try to keep new lines count to minimum

File: frontend/src/components/Register/Input.tsx [hunk 19]
```diff
@@ -0,0 +1,14 @@
+import styled from "styled-components";
+import { tertiaryColor } from "../../utils/styleVariables";
+
+
+const Input = styled.input`
+  padding: .5rem;
+  border: none;
+  border-bottom: 1px solid ${tertiaryColor};
```
Comment (ukaszcholewka) [inline frontend/src/components/Register/Input.tsx]: What if you are gonna to change website theme for blue? 
variable called `tertiaryColor` will for example be blue?

File: frontend/src/index.tsx [hunk 20]
```diff
@@ -2,11 +2,16 @@ import React from 'react';
...
 import App from './App';
+import GlobalStyles from './components/util/GlobalStyles';
+import { BrowserRouter } from 'react-router-dom';
 
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <BrowserRouter>
+      <GlobalStyles/>
```
Comment (ukaszcholewka) [inline frontend/src/index.tsx]: I think, it will be better to have `<GlobalStyles />` above router

**Review by ukaszcholewka — CHANGES_REQUESTED** [review]

**Review by ukaszcholewka — CHANGES_REQUESTED** [review]
You forgot to use `npm run lint:fix`

File: frontend/src/components/Register/Input.tsx (hunk 19 above)
Comment (GPabis) [reply frontend/src/components/Register/Input.tsx]: Then? PrimaryColor and SecoundaryColor is ok, so why TertiaryColor no?

File: frontend/src/index.tsx (hunk 20 above)
Comment (GPabis) [reply frontend/src/index.tsx]: Okk

File: frontend/src/components/Header/index.tsx (hunk 15 above)
Comment (GPabis) [reply frontend/src/components/Header/index.tsx]: Okk

File: frontend/src/components/util/ContainerLayout.tsx (hunk 16 above)
Comment (GPabis) [reply frontend/src/components/util/ContainerLayout.tsx]: Okk

File: frontend/src/components/util/GlobalStyles.tsx (hunk 17 above)
Comment (GPabis) [reply frontend/src/components/util/GlobalStyles.tsx]: Okk

File: frontend/src/components/Header/index.tsx (hunk 18 above)
Comment (GPabis) [reply frontend/src/components/Header/index.tsx]: Okk

File: backend/src/index.ts (hunk 1 above)
Comment (GPabis) [reply backend/src/index.ts]: Okk

File: backend/src/middleware/auth.ts (hunk 2 above)
Comment (GPabis) [reply backend/src/middleware/auth.ts]: Okk

File: backend/src/middleware/database.ts (hunk 4 above)
Comment (GPabis) [reply backend/src/middleware/database.ts]: okk

File: backend/src/models/user.model.ts (hunk 5 above)
Comment (GPabis) [reply backend/src/models/user.model.ts:18]: Okk

File: frontend/src/components/Header/NavElement.tsx (hunk 14 above)
Comment (GPabis) [reply frontend/src/components/Header/NavElement.tsx]: Okk

File: frontend/src/components/Register/Input.tsx (hunk 19 above)
Comment (ukaszcholewka) [reply frontend/src/components/Register/Input.tsx]: My bad, I thought ` Tertiary` is some kind of red turkey color
